darc/darc.go

```diff
@@ -1180,6 +1191,103 @@ func parseIDEvmContract(in string) (Identity, error) {
 	}, nil
 }
 
+func evalThreshold(visited map[string]bool, getDarc GetDarc,
+	attrFuncs AttrInterpreters, acceptDarc bool, s string, ids []string) error {
+
+	// we have something of form 'threshold<1/2,darc:aa,ed25519:bb,...>'
+
+	s = strings.TrimPrefix(s, "threshold<")
+	s = strings.TrimSuffix(s, ">")
+
+	entries := strings.Split(s, ",")
+
+	if len(entries) < 1 {
+		return xerrors.Errorf("expected at least the threshold fraction: %s", s)
+	}
+
+	fractionStr := entries[0]
+
+	ok := regexp.MustCompile(`\d+/\d+`).MatchString(fractionStr)
+	if !ok {
+		return xerrors.Errorf("the threshold fraction is incorrect: %s", fractionStr)
+	}
+
+	fractionElem := strings.Split(fractionStr, "/")
+	numerator, err := strconv.Atoi(fractionElem[0])
+	if err != nil {
+		return xerrors.Errorf("failed to convert numerator: %v", err)
+	}
+
+	denominator, err := strconv.Atoi(fractionElem[1])
+	if err != nil {
+		return xerrors.Errorf("failed to convert denominator: %v", err)
+	}
+
+	threshold := float64(numerator) / float64(denominator)
+
+	uniqIds := map[string]struct{}{}
+	validIds := map[string]struct{}{}
+
+	for _, entry := range entries[1:] {
+		uniqIds[entry] = struct{}{}
+
+		for _, id := range ids {
+			if id == entry {
+				validIds[id] = struct{}{}
+				break
+			}
+		}
+
+		// In case of a DARC, we need to check the _sign of that DARC and ensure
+		// we don't fall in a loop.
+		if strings.HasPrefix(entry, "darc") {
+			_, found := validIds[entry]
+			if found && acceptDarc {
+				continue
+			}
+
+			if _, ok := visited[entry]; ok {
+				return xerrors.Errorf("cycle detected")
+			}
+
+			// we make a copy so that diamond delegation will work,
+			// see TestDarc_DelegationDiamond
+			newVisited := make(map[string]bool)
+			for k, v := range visited {
+				newVisited[k] = v
+			}
+			newVisited[entry] = true
+
+			// get the latest DARC
+			d := getDarc(entry, true)
+			if d == nil {
+				continue
+			}
+
+			signExpr := d.Rules.GetSignExpr()
+			if signExpr == nil {
+				continue // no _sign on that DARC
+			}
+
+			// Recursively evaluate the sign expression of the DARC.
+			err := evalExprDarc(newVisited, signExpr, getDarc, attrFuncs, acceptDarc, ids...)
+			if err != nil {
+				continue
+			}
+
+			validIds[entry] = struct{}{}
+		}
+	}
+
+	computedFraction := float64(len(validIds)) / float64(len(uniqIds))
+	if computedFraction < threshold {
```

Comment:
Just a thought: in order to avoid floats (and possible rounding issues), you can keep the comparison in ints using:

```
 a       c
---  <  ---  <=>  a * d  <  b * c
 b       d
```

therefore coming to:

```
if len(validIDs) * denominator < len(uniqIds) * numerator {
...
```

Reply:
> I found some erronous test cases
> 
>     * `threshold<x/0,...>` yields a `NaN/Inf` threshold which is accepted
> 
>     * `threshold<x/y>` (no ids) yields a `NaN/Inf` actual threshold which is accepted

I didn't want to force users to behave correctly. If one wants to use a threshold that doesn't make sense, it is his responsibility, as long as it doesn't make the system panic.

If we want to go that path I would do the following:

A provided threshold of **0/0** is undefined,  
A provided threshold of **1/0** will always make the validation fail,  
A provided threshold of **0/N** will always make the validation pass.

Plus, we always assume that there is at least one identity in the threshold function, by flooring the computed denominator to 1.

> Also, you need to check that the denominator is equal to the number of ids.

That's not the idea. I should be able to put a threshold saying "whatever identities there are, I want a threshold of 50%", using 1/2, or 50/100, or whatever fraction.

> You should check that `numerator <= denominator` and that none are zero.

Again, we should let users do stupid things. If the denominator is greater than the numerator, that means the threshold will never be reached. Which is fine.

Reply:
Very nice! 😀